scripts/source-maintenance.sh

```diff
@@ -117,7 +133,12 @@ if test "${ASVER}" != "${TargetAstyleVersion}" ; then
 		echo "Formatting anyway, please double check output before submitting"
 	fi
 else
-	echo "Found astyle ${ASVER}. Formatting..."
+	echo "Found astyle ${ASVER}"
+fi
+
+if [ ! -f src/http/Makefile ]; then
+    echo "please run ./bootstrap.sh && ./configure to prepare xperf sources"
```

Comment:
What about developers (like me) who build out of tree to keep automated junk away from their source checkouts?

Reply:
@yadij, I assume you question the `-f` check rather than the echoed advice, and that you normally run the script from the checkout directory rather than in the build directory. Does the existing `make -C src/http gperf-files` command in the official script work for your out-of-tree builds? If it does, then the above check should work as well, right? If the official script does not work, then I would suggest fixing this problem in a dedicated PR (and removing this new check from this PR).

@kinkie, this is one of those places where `--quick-update-since` or a similar "smart" option may work better because it can exclude gperf code regardless of whether (and where) the sources were ./configured.

Another wrinkle here is that while we (will eventually) want all developers to run our script so that their PRs are properly prepped, most developers should probably not be _forced_ to run gperf because automated gperf application
* is gperf version-dependent (but there are no corresponding safeguards),
* sometimes produces completely bogus results, _and_
* is actually unnecessary in most cases.

It feels like we should apply the same logic to astyle and gperf:
1. By default, apply each (with a version check) and quit on failures. The error message can provide a reference to detailed solution instructions. Some developers will then install the right missing tools and achieve full formatting. Some will _knowingly_ bypass the step instead. This is the best _default_ I can think of.
2. Add an option to explicitly bypass each step. For example, `--without-astyle` and `--without-gperf`.
3. Bypass the version check when an explicit `--with-...` option is given, with a warning. This will let folks to check what difference their tool version makes (without editing the script).

If there is agreement regarding these or similar enhancements, they can be done in this or a dedicated PR.

Reply:
> Does the existing make -C src/http gperf-files command in the official script work for your out-of-tree builds?

No it does not. I think now we may need to be able to specify a --builddir option (defaults to '.') to use in "test ! -f $builddir/src/Makefile" and "make -C $builddir/src/http gperf-files" commands.

OR,
 maybe it is time to fold astyle and similar maintenance actions into Makefile.in logic on a custom target like gperf uses. That way the CI can highlight issues during pr-test checking instead of leaving it to QA reviewers or maintenance commits.

Reply:
Regenerating xperf files is a rare occurrence, isn't it? I've iterated by skipping the step and emitting a warning instead

Reply:
> Amos: I think now we may need to be able to specify a --builddir option (defaults to '.') to use in "test ! -f $builddir/src/Makefile" and "make -C $builddir/src/http gperf-files" commands.

No option should be needed to trigger out-of-tree maintenance mode AFAICT: One simply runs the script from the build directory instead of the source directory. Needless to say, we will need to do more to _support_ out-of-tree maintenance. I wish our collective energy would be spent elsewhere, but since the Project seems to support (in general) both in-tree and out-of-tree builds, I see no reason why such improvements should be rejected per se.

> maybe it is time to fold astyle and similar maintenance actions into Makefile.in logic on a custom target like gperf uses.

I think that is an orthogonal improvement. Whether or not we add new "make foo" targets for source maintenance, those targets will just start the source-maintenance.sh script (possibly with options) or equivalent -- we do _not_ want to write non-trivial code _inside_ Makefiles! Thus, to support out-of-tree maintenance, we would need to implement out-of-tree maintenance changes _inside_ our scripts, regardless of whether that maintenance is Makefile-activated or not.


> That way the CI can highlight issues during pr-test checking instead of leaving it to QA reviewers or maintenance commits.

This last bit feels irrelevant: Semaphore CI already detects (some) source maintenance issues and there is no reason it (or Jenkins) cannot be enhanced to detect all of them -- in fact, such full detection is the long term goal! One does not need to be outside the source tree to (build Squid and) detect these issues.